### DISCUSSION TITLE
Problems with appended attributes.

### DISCUSSION THREAD

**Discussion by pacorado** [livewire/livewire#3836] [1. Help] [answered]
Hi,

First of all: I love Livewire. But I think I found an issue. 

I have a model "Profile" with an appended attribute. This model is stored in the session. When I fetch the model from the session in the mount method of my Livewire component, modify it in another method of the Livewire component and then store it in the session, the appended attribute is not serialized. When I retrieve the Profile model from the session in the method that modifies it everything works as expected.

I am not sure if it is my lack of understanding or a bug in Livewire but find it curious. I hope Caleb or somebody else can help me with this. In the uploaded ZIP file I have included the source files with the only the relevant methods.

Best regards.

[AppendedAttributesNotWorking.zip](https://github.com/livewire/livewire/files/7206837/AppendedAttributesNotWorking.zip)

**Comment by pacorado** [comment]
Code blocks for ease of reading:



**\App\Models\Profile.php**

```php

<?php



namespace App\Models;



use Carbon\Carbon;
use Illuminate\Database\Eloquent\Model;



/**
 * @mixin IdeHelperProfile
 */
class Profile extends Model
{



    /**
     * The key within the session array for the selected profile.
     *
     * @const string
     */
    public const SESSION_VARIABLE_NAME = 'selected_profile';



    /**
     * Gets the previously stored - or a blank - profile from the session.
     *
     * @return static
     */
    public static function selected(): static
    {
        return session()->has(static::SESSION_VARIABLE_NAME)
            ? session()->get(static::SESSION_VARIABLE_NAME)
            : tap(new static(), fn(self $profile) => $profile->toSession());
    }



    /**
     * The attributes appended to this model.
     *
     * @var array
     */
    protected array $appendedAttributes = [
        'online_since' => null,
    ];



    /**
     * The accessors to append to the model's array form.
     *
     * @var array
     */
    protected $appends = [
        'online_since',
    ];



    /**
     * The model's attributes.
     *
     * @var array
     */
    protected $attributes = [
        'show_real_name' => false,
    ];



    /**
     * Gets the date and time this profile came online.
     *
     * @return Carbon|null
     */
    public function getOnlineSinceAttribute(): Carbon|null
    {
        return empty($this->appendedAttributes['online_since'])
            ? null
            : $this->asDateTime($this->appendedAttributes['online_since']);
    }



    /**
     * Gets the date and time this profile came online.
     *
     * @param Carbon|string|null $online_since
     *
     * @return void
     */
    public function setOnlineSinceAttribute(Carbon|string|null $online_since): void
    {
        $this->appendedAttributes['online_since'] = empty($online_since)
            ? null
            : $this->fromDateTime($online_since);
    }



    /**
     * Stores this profile in the session.
     *
     * @returm void
     */
    public function toSession(): void
    {
        session()->put(static::SESSION_VARIABLE_NAME, $this);
    }



}

```

**Working version of \App\Http\Livewire\Web\Dashboard\Index\InterfaceLanguageSelector**
```php
<?php

namespace App\Http\Livewire\Web\Dashboard\Index;

use App\Actions\Auth\ChangeInterfaceLanguageAction;
use App\Models\Language;
use App\Models\Profile;
use DB;
use Exception;
use Illuminate\Database\Eloquent\Collection;
use Livewire\Component;
use Throwable;

class InterfaceLanguageSelector extends Component
{

    public Collection $languages;
    public string $interface_language_locale;



    public function mount()
    {
        $this->languages = Language::orderBy('locale')->get();
        $this->interface_language_locale = app()->getLocale();
    }



    public function render()
    {
        return view('livewire.web.dashboard.index.interface-language-selector');
    }



    /**
     * @throws Throwable
     */
    public function setInterfaceLanguageLocale(string $interface_language_locale): void
    {
        $this->validate();

        DB::beginTransaction();
        try {
            $profile = Profile::selected();
            app(ChangeInterfaceLanguageAction::class)->execute($profile, $interface_language_locale);
            DB::commit();
            $profile->toSession();
            $this->interface_language_locale = $interface_language_locale;
            $this->emit('livewire-document-reload');
        } catch (Exception $e) {
            DB::rollBack();
        }
    }



    public function submit(): void
    {
    }



    protected function rules(): array
    {
        return [
            'interface_language_locale' => [
                'string',
                'required',
                'exists:languages,locale,is_interface_language,TRUE'
            ],
        ];
    }

}
```

**NOT Working version of \App\Http\Livewire\Web\Dashboard\Index\InterfaceLanguageSelector**
```php
<?php

namespace App\Http\Livewire\Web\Dashboard\Index;

use App\Actions\Auth\ChangeInterfaceLanguageAction;
use App\Models\Language;
use App\Models\Profile;
use DB;
use Exception;
use Illuminate\Database\Eloquent\Collection;
use Livewire\Component;
use Throwable;

class InterfaceLanguageSelector extends Component
{

    public Profile $profile;
    public Collection $languages;
    public string $interface_language_locale;



    public function mount()
    {
        $this->profile = Profile::selected();
        $this->languages = Language::orderBy('locale')->get();
        $this->interface_language_locale = app()->getLocale();
    }



    public function render()
    {
        return view('livewire.web.dashboard.index.interface-language-selector');
    }



    /**
     * @throws Throwable
     */
    public function setInterfaceLanguageLocale(string $interface_language_locale): void
    {
        $this->validate();

        DB::beginTransaction();
        try {
            app(ChangeInterfaceLanguageAction::class)->execute($this->profile, $interface_language_locale);
            DB::commit();
            $this->profile->toSession();
            $this->interface_language_locale = $interface_language_locale;
            $this->emit('livewire-document-reload');
        } catch (Exception $e) {
            DB::rollBack();
        }
    }



    public function submit(): void
    {
    }



    protected function rules(): array
    {
        return [
            'interface_language_locale' => [
                'string',
                'required',
                'exists:languages,locale,is_interface_language,TRUE'
            ],
        ];
    }

}

```

**Comment by joshhanley** [comment]
@pacorado thanks! Pasting code is much easier for us to review than a zip file 🙂 I have edited your comment to add type hinting to the code blocks so it styles the code properly.

Can you confirm what patch version of Livewire you are using?

**Comment by pacorado** [comment]
Hi @joshhanley, As I am a solo dev I have the luxury of running "npm update && composer u" each morning without upsetting others. So I use the latest and greatest. My current version of Livewire 2.6.5. Thank you for asking. I will also turn on code hinting in future questions. I learn something everyday!

**Reply by joshhanley** [reply]
@pacorado ok great, thanks for confirming 🙂 hmm ok we will have to investigate to see what is going on

**Comment by joshhanley** [comment, marked as answer]
@pacorado I have just run a simplified test locally, to see if appended data is being pushed and restored from the session correctly, and it is for me, so I suspect there might be an issue with your appends attributes mutator logic, I couldn't get that part working correctly.

But I don't think that is a Livewire issue 🙂

I'd recommend testing it out manually (without Livewire), pushing to the session using tinker/tinkerwell and seeing what happens.

Hope this helps!

**Reply by pacorado** [reply]
Thank you for checking it out. I also have no idea what is going on. The method "fromDateTime" is a Laravel method defined in the "HasAttributes" trait and just takes care of the casting. There is no further logic. (The action just accepts the parameters, set it in the model and calls the save method.) 

For now I just refactored my code to retrieve the data from the session as needed. First of all did I forget that the public attributes of a Livewire component are exposed and I do not want the world to be able to look at the Profile model. Second of all there are more Livewire components on the dashboard that can modify the Profile model and that can become problematic also.

So again, thank you for your efforts in this matter and all other code you added to Livewire. (I just read the article about 2.6 in Laravel News and saw your name there) It is really REALLY appreciated. I think Livewire (and AlpineJS) are great!
